File: back/api/src/database/repository/room.py
from src.database.connection_db import Database
import json
import uuid
from bson import json_util 

class RoomDAO: # DAO - Data Access Object
    def __init__(self):
        self.db = Database(collection="room")

    def get_all(self):
        try:
            res = self.db.collection.find()

            parsed_json = json.loads(json_util.dumps(res))
            return parsed_json
        
        except Exception as e:
            print(f"Houve um erro ao tentar pegar todas as salas: {e}")
            return None
    
    def create(self, new_room_name):
        try:
            uuid_room = str(uuid.uuid4())

            # if self.read_one(new_room_name) != (None and True):
            #     return {"Error message": "Sala já existente"}

            
            room_json = {"name": new_room_name, "uuid": uuid_room}
            res = self.db.collection.insert_one(room_json)
            
            return True
        except Exception as e:
            print(f"Houve um erro ao tentar criar uma nova sala: {e}")
            return False

    def read_one(self, name):
        try:
            res = self.db.collection.find_one({"name": name})
            print("one room: ", res)

            parsed_json = json.loads(json_util.dumps(res))

            return parsed_json
        except Exception as e:
            return False
        
    def delete(self, uuid):
        try:
            res = self.db.collection.delete_one({"uuid": uuid})

            if res.deleted_count == 0:
                return False
            else:
                return True
        except Exception as e:
            print(f"Houve um erro ao tentar apagar a sala:  {e}")
            return None
    
    def update_equipment_in_room(self, name_room, name_equipment, patrimonio):
        try:
            new_data = {
                "equipment": name_equipment,
                "patrimonio": patrimonio,
            }
            res = self.db.collection.update_one({"name": name_room}, {"$push": {"equipments": new_data}})
            print("one equipment: ", res)

            return res
        except Exception as e:
            raise e
        
    def delete_equipment_in_room(self, patrimonio):
        try:
            res = self.db.collection.update_many({}, {"$pull": {"equipments": {"patrimonio": patrimonio}}})
            # print("one equipment: ", res)

            return res
        except Exception as e:
            raise e

File: front/application/lib/database/models/Room.dart
class Room {
  Id? iId;
  String? name;
  List<Equipments>? equipments;

  Room({this.iId, this.name, this.equipments});

  Room.fromJson(Map<String, dynamic> json) {
    iId = json['_id'] != null ? new Id.fromJson(json['_id']) : null;
    name = json['name'];
    if (json['equipments'] != null) {
      equipments = <Equipments>[];
      json['equipments'].forEach((v) {
        equipments!.add(new Equipments.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    if (this.iId != null) {
      data['_id'] = this.iId!.toJson();
    }
    data['name'] = this.name;
    if (this.equipments != null) {
      data['equipments'] = this.equipments!.map((v) => v.toJson()).toList();
    }
    return data;
  }
}

class Id {
  String? oid;

  Id({this.oid});

  Id.fromJson(Map<String, dynamic> json) {
    oid = json['$oid'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['$oid'] = this.oid;
    return data;
  }
}

class Equipments {
  String? equipment;
  String? patrimonio;

  Equipments({this.equipment, this.patrimonio});

  Equipments.fromJson(Map<String, dynamic> json) {
    equipment = json['equipment'];
    patrimonio = json['patrimonio'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['equipment'] = this.equipment;
    data['patrimonio'] = this.patrimonio;
    return data;
  }
}
